HttpListener - Prefix WWW-Authenticate header with scheme

mattpwhite opened the pull request:
Fixes https://github.com/dotnet/corefx/issues/26606

rmkerr commented:
Glad to see you got this fix together. The spec is super clear that this is the expected behavior:
```
   The negotiate scheme will operate as follows:
           challenge       = "Negotiate" auth-data
           auth-data       = 1#( [gssapi-data] )
```
I'm going to run the outerloop tests against the change, and if it passes the change looks good to me. Thanks for taking this fix on @mattpwhite!

@dotnet-bot test OuterLoop Windows x64 Debug Build please
@dotnet-bot test OuterLoop Linux x64 Debug Build please
@dotnet-bot test OuterLoop OSX x64 Debug Build please

wfurt commented:
should we add test for this?

mattpwhite commented:
> should we add test for this?

If you have a way to have a functional Kerberos infrastructure in a test environment...  There was some discussion of this in https://github.com/dotnet/corefx/issues/26606, and it seemed like the answer was no and manual testing would have to suffice.  The code I used to test is here - https://github.com/mattpwhite/httplistener-examples.  That code can also be used to demonstrate that HttpWebRequest has a bug where it claims to do mutual auth, but in fact does no such thing.  That wasn't my actual problem, so I didn't pursue it.

davidsh commented:
@Tratcher PTAL

mattpwhite commented:
Those test failures seem unrelated.  Unless I'm missing something obvious, none of them have any relation to HttpListener.

Also, questing about the CLA.  I have approval from my employer for this, but if you need that signed I have to have someone further up the chain from me actually sign it.  I'm just an engineer and don't have the authority to make some of the assertions in there on my employer's behalf (or so legal tells me).  Someone with that authority is willing to sign for me in relation to this GitHub account.  Does that work?

Tratcher commented on src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs:
```diff
@@ -1150,6 +1150,12 @@ internal HttpListenerContext HandleAuthentication(RequestContextBase memoryBlob,
                         }
 
+                        // Prefix SPNEGO token with Negotiate scheme per RFC 4559
+                        if (!String.IsNullOrEmpty(outBlob) && headerScheme == AuthenticationSchemes.Negotiate)
```
Does outblob ever have a value at this stage for protocols besides Negotiate? It doesn't look like it. Move this up into the negotiate code block. See:
https://github.com/mattpwhite/corefx/blob/3ef8e6cd59787df1859e62bd69518938db8127c8/src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs#L1011
https://github.com/mattpwhite/corefx/blob/3ef8e6cd59787df1859e62bd69518938db8127c8/src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs#L1103-L1107

mattpwhite replied:
Maybe, I'll have a look.  This method is umm... sprawling.  It has been a little bit since I actually wrote this, so I'll need a bit to re-read everything.

mattpwhite replied:
I think we only end up in that branch at 1100 if we haven't completed establishing the context (meaning we're going to 401 and send an NTLM challenge or we need to exchange another round of SPNEGO blobs for negotiate).  In the mutual auth case, `context.IsCompleted` should be true (it is from the server's perspective) we go down the other path at 1018.  I guess it could go right there, is that what you're suggesting?

Tratcher replied:
Yes, right around 1011.

mattpwhite commented:
> Nice find. You said AspNetCore HttpSys works fine, correct? 

Yeah, the kernel will insert the correct header unless user mode has already set it.  HttpSysListener doesn't set the header explicitly, so it was fine.

karelz commented:
Regarding CLA - I see "_All CLA requirements met._" in the results list on this PR. All should be fine. I believe that for small changes (few lines like this one), there is no CLA signature required, the CLA signature is required only for larger changes - see https://cla.dotnetfoundation.org or https://cla.dotnetfoundation.org/dotnet/corefx for details.
If you ever contribute larger change, you need to sign the CLA (the bot will ask for it) - if you hit troubles, I can point you in the right direction.

mattpwhite commented:
Sorry for the delay, I moved the prefixing up as requested, and removed prefixing from the path that generates the challenge since we now do it unconditionally earlier.  Let me know if that's what you wanted.

davidsh commented on src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs:
```diff
@@ -1008,7 +1008,10 @@ internal HttpListenerContext HandleAuthentication(RequestContextBase memoryBlob,
                             {
-                                outBlob = Convert.ToBase64String(decodedOutgoingBlob);
+                                // Prefix SPNEGO token/NTLM callenge with scheme per RFC 4559, MS-NTHT
+                                outBlob = String.Format("{0} {1}",
```
CoreFx coding guidelines use 'string' instead of "String".

davidsh commented on src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs:
```diff
@@ -1099,12 +1102,9 @@ internal HttpListenerContext HandleAuthentication(RequestContextBase memoryBlob,
-                                    {
-                                        challenge += " " + outBlob;
-                                    }
+                                    challenge = String.IsNullOrEmpty(outBlob)
```
CoreFx coding guidelines use 'string' instead of "String".

mattpwhite commented:
Changed to `string`.  I used `String` because that's what was used there before my changes.

davidsh commented:
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

mattpwhite commented:
Anything left for me to do here?  Just making sure I’m not missing something.

rmkerr commented:
Thanks for pinging us on this! The test failures look unrelated, but I'm going to re-run them anyways. Pending those passing, we can get this merged.

@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

mattpwhite commented:
Another gentle ping.  I had hoped this would make 2.1, has that ship already sailed?

rmkerr commented:
All of the Linux test runs are showing catastrophic failures in the http functional tests, but in the actual logs they appear to be passing:
```
2018-03-22 05:13:15,589: INFO: proc(54): run_and_log_output: Output: === TEST EXECUTION SUMMARY ===
2018-03-22 05:13:15,602: INFO: proc(54): run_and_log_output: Output:    System.Net.Http.Functional.Tests  Total: 1309, Errors: 0, Failed: 0, Skipped: 24, Time: 35.388s
2018-03-22 05:13:15,677: INFO: proc(54): run_and_log_output: Output: ~/dotnetbuild/work/e70a5406-781d-4185-a738-96cf938ce2e3/Work/0657be07-3932-436e-8bcc-84d1f6bb8029/Unzip
```
Since the tests look good, I think that this is okay to merge.

rmkerr commented:
Actually after talking to others, it is probably a good idea to be sure that the test failures are unrelated. Running them again:

@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Windows x86 Release Build

karelz commented:
@rmkerr can you please help with the test failures. We may be able to get it in 2.1 today/tomorrow if the test results are clean.

rmkerr commented:
The http test failure on Debian 9 was caused by dumpling -- I can verify in the logs that the tests actually passed:
```
2018-03-23 18:26:24,938: INFO: proc(54): run_and_log_output: Output: === TEST EXECUTION SUMMARY ===
2018-03-23 18:26:24,948: INFO: proc(54): run_and_log_output: Output:    System.Net.Http.Functional.Tests  Total: 2027, Errors: 0, Failed: 0, Skipped: 26, Time: 114.247s
2018-03-23 18:26:25,052: INFO: proc(54): run_and_log_output: Output: ~/dotnetbuild/work/03dc50b4-1c24-43eb-8205-e2ca5f589ffe/Work/eb57bf12-fea3-4a2c-9613-640b5f918c18/Unzip
2018-03-23 18:26:25,053: INFO: proc(54): run_and_log_output: Output: ----- end 18:26:25 ----- exit code 0 ----------------------------------------------------------
```

rmkerr commented:
@dotnet-bot test OuterLoop OSX x64 Debug Build please

rmkerr commented:
Merged. Thanks for making this fix @mattpwhite!

mattpwhite commented:
Thanks!
